Mathlib/RingTheory/Polynomial/DegreeLT.lean

```diff
@@ -0,0 +1,238 @@
+    simp [WithBot.bot_lt_iff_ne_bot, ne_bot_of_gt hp]
+  · exact (degree_mul_le _ _).trans_lt (WithBot.add_lt_add_of_lt_of_le hq0 hp hq)
+
+@[aesop unsafe 50%]
+theorem degree_mul_lt_of_lt_right {p q : R[X]} {a : ℕ} {b : WithBot ℕ}
+    (hp : degree p ≤ a) (hq : degree q < b) :
+    degree (p * q) < a + b := by
+  by_cases hp0 : p.degree = ⊥
+  · rw [degree_eq_bot] at hp0
+    rw [hp0, zero_mul, degree_zero, WithBot.bot_lt_add]
+    simp [WithBot.bot_lt_iff_ne_bot, ne_bot_of_gt hq]
+  · exact (degree_mul_le _ _).trans_lt (WithBot.add_lt_add_of_le_of_lt hp0 hp hq)
+
+lemma mul_left_mem_degreeLT {n} (p q : R[X]) (hp : degree p ≤ m) (hq : q ∈ R[x]_n) :
+    p * q ∈ R[x]_(m + n) := by
+  rw [mem_degreeLT] at *
+  exact degree_mul_lt_of_lt_right hp hq
+
+lemma mul_left_mem_degreeLT' {n} (p q : R[X]) (hp : degree p ≤ m) (hq : q ∈ R[x]_n) :
+    p * q ∈ R[x]_(n + m) := by
+  rw [add_comm]
+  exact mul_left_mem_degreeLT _ _ ‹_› ‹_›
+
+lemma mul_right_mem_degreeLT {m} (p q : R[X]) (hp : p ∈ R[x]_m) (hq : degree q ≤ n) :
+    p * q ∈ R[x]_(m + n) := by
+  rw [mem_degreeLT] at *
+  exact degree_mul_lt_of_lt_left hp hq
+
+lemma mul_right_mem_degreeLT' {m} (p q : R[X]) (hp : p ∈ R[x]_m) (hq : degree q ≤ n) :
+    p * q ∈ R[x]_(n + m) := by
+  rw [add_comm]
+  exact mul_right_mem_degreeLT _ _ ‹_› ‹_›
+
+end degreeLT
+
+section translate
+
+variable {R : Type*} [CommRing R] {x : R} {m n : ℕ} {a : R} {f g : R[X]}
+
+variable (x) in
+/-- TO MOVE? An automorphism of `R[X]` given by a substitution of the form `X ↦ X - a`. -/
+noncomputable def translate : R[X] ≃+* R[X] where
+  invFun    := (.X - .C (-x) : R[X]).compRingHom
+  left_inv  := fun P ↦ by simp [comp_assoc]
+  right_inv := fun P ↦ by simp [comp_assoc]
+  __ := (.X - .C x : R[X]).compRingHom
+
+lemma translate_apply : f.translate x = f.comp (.X - .C x) :=
+  rfl
+
+@[simp] lemma translate_symm_apply : (translate x).symm f = f.translate (-x) :=
+  rfl
+
+@[simp] lemma translate_C : (C a).translate x = C a :=
+  C_comp
+
+@[simp] lemma translate_smul : (a • f).translate x = a • f.translate x := by
+  rw [smul_eq_C_mul, smul_eq_C_mul, map_mul, translate_C]
+
+lemma natDegree_translate_le : (f.translate x).natDegree ≤ f.natDegree :=
+  natDegree_comp_le.trans <| by nontriviality R; rw [natDegree_X_sub_C, mul_one]
+
+lemma natDegree_translate : (f.translate x).natDegree = f.natDegree := by
+  refine eq_of_le_of_le natDegree_translate_le ?_
+  conv_lhs => rw [← (translate x).symm_apply_apply f, translate_symm_apply]
+  exact natDegree_translate_le
+
+lemma degree_translate : (f.translate x).degree = f.degree := by
+  by_cases h : f = 0
+  · rw [h, map_zero]
+  · rw [degree_eq_natDegree h, degree_eq_natDegree ((translate x).map_ne_zero_iff.2 h),
+      natDegree_translate]
+
+variable (x n) in
+/-- The `translate` map induces an automorphism of the module `R[x]_n` of polynomials of
+degree `< n`. -/
+noncomputable def translateLinear : (R[x]_n) ≃ₗ[R] (R[x]_n) where
```

Comment:
Again I think the code feels easier to read if this was written `translateLinear (a : R) : ...` but this is only a minor thing.